src/contracts/reader/GPv2TradeSimulator.sol

```diff
@@ -0,0 +1,135 @@
+    /// order. This makes using this method inaccurate for predicting gas usage
+    /// for specific trades. However, seeing as computing trade executions is
+    /// completely independant of external contracts (i.e. it does not depend on
+    /// the tokens being traded or the interactions being executed) the gas
+    /// consumption should be off by a predicatable amount.
+    ///
+    /// @param trade The trade to simulate.
+    /// @param interactions A set of interactions to settle the trade against.
+    function simulateTrade(
+        Trade calldata trade,
+        GPv2Interaction.Data[] calldata interactions
+    ) external returns (int256 sellDelta, int256 buyDelta) {
+        GPv2Settlement self = GPv2Settlement(payable(address(this)));
+        Context memory context = createContext(trade);
+
+        self.vaultRelayer().transferFromAccounts(context.inTransfers);
+
+        for (uint256 i; i < interactions.length; i++) {
+            GPv2Interaction.execute(interactions[i]);
+        }
+
+        updateOutTransferAmount(context);
+        self.vault().transferToAccounts(context.outTransfers);
+
+        sellDelta = computeBalanceDelta(context.sellTokenState);
+        buyDelta = computeBalanceDelta(context.buyTokenState);
+    }
+
+    /// @dev Initializes a simulation context in memory for the current trade
+    /// simulation.
+    ///
+    /// This helps organize simulation data and work around "stack too deep"
+    /// Solidity errors.
+    function createContext(Trade calldata trade)
+        private
+        view
+        returns (Context memory context)
+    {
+        context.inTransfers = new GPv2Transfer.Data[](1);
+        context.inTransfers[0].account = trade.owner;
+        context.inTransfers[0].token = trade.sellToken;
+        context.inTransfers[0].amount = trade.sellAmount;
+        context.inTransfers[0].balance = trade.sellTokenBalance;
+
+        context.outTransfers = new GPv2Transfer.Data[](1);
+        context.outTransfers[0].account = trade.receiver ==
+            GPv2Order.RECEIVER_SAME_AS_OWNER
+            ? trade.owner
+            : trade.receiver;
+        context.outTransfers[0].token = trade.buyToken;
+        context.outTransfers[0].amount = trade.buyAmount;
+        context.outTransfers[0].balance = trade.buyTokenBalance;
+
+        context.sellTokenState.token = trade.sellToken;
+        context.sellTokenState.initialBalance = trade.sellToken.balanceOf(
+            address(this)
+        );
+
+        context.buyTokenState.token = trade.buyToken;
+        context.buyTokenState.initialBalance = trade.buyToken.balanceOf(
+            address(this)
+        );
+    }
+
+    /// @dev Updates the out transfer token amount to be the exact amount of buy
+    /// token that has been received so far if the trade simulation was done
+    /// using a special sentinal values for the buy amount.
+    function updateOutTransferAmount(Context memory context) private view {
```

Comment:
What are the use cases where we would want to use that vs a fixed amount?

Reply:
Hmm, now that I think of it, we probably don't want more information to be returned because, as its currently implemented we can let through buy tokens where we don't transfer the full amount.

Reply:
Maybe we can actually use this to still quote the user prices for such tokens (by taking into account that the last transfer will lose some tokens due to fee)

Reply:
I've updated the simulation result to include more information. Specifically:
- `gasUsed`: The rational behind this is to be able to get the result of the simulation as well as a rough gas estimate in a single RPC call (instead of relying on a pair of `eth_estimateGas` and `eth_call`).
- `executedBuyAmount`: This is useful in conjunction with the `0` buy amount in the trade simulation parameter, to record exactly what was transferred.
- `ownerBalance`: Instead of just computing the balance deltas for the settlement contract, we also compute the balance deltas for the trade's `owner` account. This allows to do checks like `ownerBalance.{sell,buy}TokenDelta + contractBalance.{sell,buy}TokenDelta == 0` not only does the contract receive the full amount transferred in (+/- 1 for tokens like `stETH`) but that the trader also receives the full amount transferred out.

This should give us quite a bit of flexibility when it comes to flagging unsupported tokens, while still keeping the same rough functionality when it comes to estimating gas costs.

All in all, if we deploy this, I figure its better to err on the side of caution and include more features to avoid requiring re-deploys.